src/library/scala/runtime/ScalaRunTime.scala

```diff
@@ -248,30 +376,63 @@ object ScalaRunTime {
     def inner(arg: Any): String = arg match {
       case null                         => "null"
       case ""                           => "\"\""
-      case x: String                    => if (x.head.isWhitespace || x.last.isWhitespace) "\"" + x + "\"" else x
-      case x if useOwnToString(x)       => x.toString
+      case s: String if verboseProduct  => stringToString(s)
+      case x: String                    => if (x.head.isWhitespace || x.last.isWhitespace) s""""$x"""" else x
       case x: AnyRef if isArray(x)      => arrayToString(x)
-      case x: scala.collection.Map[_, _] => x.iterator.take(maxElements).map(mapInner).mkString(x.collectionClassName + "(", ", ", ")")
-      case x: Iterable[_]               => x.iterator.take(maxElements).map(inner).mkString(x.collectionClassName + "(", ", ", ")")
-      case x: Product1[_] if isTuple(x) => "(" + inner(x._1) + ",)" // that special trailing comma
-      case x: Product if isTuple(x)     => x.productIterator.map(inner).mkString("(", ",", ")")
+      case
+           // Range/NumericRange have a custom toString to avoid walking a gazillion elements
+           _: Range | _: NumericRange[_]
+           // StringBuilder(a, b, c) and similar not so attractive
+         | _: StringView | _: StringOps | _: StringBuilder
+           // Don't want to evaluate any elements in a view
+         | _: View[_]                   => arg.toString
+      // Use toString if not eagerly constructed, to avoid traversing infinity or forcing lazy elements.
+      // Custom collections are deemed to have useful toString (scala/bug#3710): either they override className
+      // to leverage default toString (mkString) which is always mixed-in, or they override toString or possibly addString.
+      // xml.Node and xml.MetaData return themselves in their iterator, so use their xml toString instead.
+      // Among other unsavory types, AbstractFile is Iterable[AbstractFile] but not strict, so it will use toString to show name.
+      case x: AbstractIterable[_] if x.isTraversableAgain && (
+             !x.isInstanceOf[StrictOptimizedIterableOps[_, AnyConstr, _]] ||
+             !isScalaClass(x) && commonCollectionNames.contains(x.collectionClassName) ||
```

Comment:
Maybe this is too magic for users?

```
scala> class K extends Seq[Int] { def iterator = Iterator(1); def apply(i: Int) = 1; def length: Int = 1; override def toString = "ka" }
class K

scala> new K
val res0: K = Seq(1) // used to be ka
```

Reply:
It's probably not too bizarre for REPL to look at the byte code for toString to detect if it is trustworthy (not just the mixin that might hang). I wasn't satisfied with this heuristic/guesswork.

Reply:
Oh it's the change to test for `AbstractIterable` instead of `Iterable`. What is the rationale for that?

Reply:
That idea was a non-trivial impl will extend it.

Reply:
I'd say let's stay with `Iterable`. The special treatment for collections below matches `Iterable`, and the idea here is to narrow down that custom stringification to "known" types.

Then this PR seems good to go.

Reply:
The idea (from the ticket) was to avoid printing an infinite `Iterator.continually` or whatever. It seemed like REPL could have a simple criterion for when to take over string duties. The user story is that the user is noodling around in REPL. I'll give it a bit more thought or just try it out.